core/chains/evm/logpoller/log_poller.go

```diff
@@ -707,23 +707,72 @@ func (lp *logPoller) LatestLogEventSigsAddrsWithConfs(fromBlock int64, eventSigs
 	return lp.orm.SelectLatestLogEventSigsAddrsWithConfs(fromBlock, addresses, eventSigs, confs, qopts...)
 }
 
-// GetBlocks tries to get the specified block numbers from the log pollers
-// blocks table. Returns the blocks it was able to find, empty slice if none.
-// When the log poller does not have requested blocks, it falls back
-// to RPC to fetch the missing blocks.
-// response contains blocks in the same order as "numbers" in request parameters
-// the first context parameter takes precedence over contexts passed through qopts
-func (lp *logPoller) GetBlocks(ctx context.Context, numbers []uint64, qopts ...pg.QOpt) ([]LogPollerBlock, error) {
+// GetBlocksRange tries to get the specified block numbers from the log pollers
+// blocks table. It falls back to the RPC for any unfulfilled requested blocks.
+func (lp *logPoller) GetBlocksRange(ctx context.Context, numbers []uint64, qopts ...pg.QOpt) ([]LogPollerBlock, error) {
+	var blocks []LogPollerBlock
+
+	// Do nothing if no blocks are requested.
+	if len(numbers) == 0 {
+		return blocks, nil
+	}
+
+	// Sort the a copy of the requested blocks in ascending order.
+	sortedNumbers := make([]uint64, len(numbers))
+	sort.Slice(sortedNumbers, func(i, j int) bool { return numbers[i] < numbers[j] })
+
+	// Assign the requested blocks to a mapping.
+	blocksRequested := make(map[uint64]struct{})
+	for _, b := range numbers {
+		blocksRequested[b] = struct{}{}
+	}
+
+	// Retrieve all blocks within this range from the log poller.
 	blocksFound := make(map[uint64]LogPollerBlock)
 	qopts = append(qopts, pg.WithParentCtx(ctx))
-	lpBlocks, err := lp.orm.GetBlocks(numbers, qopts...)
+	lpBlocks, err := lp.orm.GetBlocksRange(sortedNumbers[0], sortedNumbers[len(numbers)-1], qopts...)
 	if err != nil {
 		lp.lggr.Warnw("Error while retrieving blocks from log pollers blocks table. Falling back to RPC...", "requestedBlocks", numbers, "err", err)
+	} else {
+		for _, b := range lpBlocks {
+			if _, ok := blocksRequested[uint64(b.BlockNumber)]; ok {
+				// Only fill requested blocks.
+				blocksFound[uint64(b.BlockNumber)] = b
+			}
+		}
+		lp.lggr.Debugw("Got blocks from log poller", "blockNumbers", maps.Keys(blocksFound))
+	}
+
+	// Fill any remaining blocks from the client.
+	blocksFoundFromRPC, err := lp.fillRemainingBlocksFromRPC(ctx, blocksFound, numbers)
+	if err != nil {
+		return nil, err
+	}
+	for num, b := range blocksFoundFromRPC {
+		blocksFound[num] = b
+	}
+
+	var blocksNotFound []uint64
+	for _, num := range numbers {
+		b, ok := blocksFound[num]
+		if !ok {
+			blocksNotFound = append(blocksNotFound, num)
+		}
+		blocks = append(blocks, b)
 	}
-	for _, b := range lpBlocks {
-		blocksFound[uint64(b.BlockNumber)] = b
+
+	if len(blocksNotFound) > 0 {
+		return nil, errors.Errorf("blocks were not found in db or RPC call: %v", blocksNotFound)
 	}
 
+	return blocks, nil
+}
+
+func (lp *logPoller) fillRemainingBlocksFromRPC(
+	ctx context.Context,
+	blocksFound map[uint64]LogPollerBlock,
```

Comment:
Don't need to pass this in anymore right?

Reply:
we still compare blocksFound to blocknumbers to make the RPC call.

Reply:
I could do this logic outside of this check to make it more intuitive, if that is better.

Reply:
Renamed the params to make this a bit more intuitive.

Reply:
Ah, got it. The diff was just a bit tricky to read.